core/libp2p/multi/utils/protocol_list.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_PROTOCOLLIST_HPP
+#define KAGOME_PROTOCOLLIST_HPP
+
+#include <functional>
+#include <map>
+#include <array>
+
+namespace libp2p::multi {
+
+  /**
+   * Contains some data about a network protocol, e. g. its name and code
+   */
+  struct Protocol {
+   public:
+    /**
+     * Denotes that the size of the protocol is variable
+     */
+    static const int kVarLen = -1;
+
+    enum class Code : std::size_t {
+      ip4 = 4,
+      tcp = 6,
+      udp = 273,
+      dccp = 33,
+      ip6 = 41,
+      ip6zone = 42,
+      dns = 53,
+      dns4 = 54,
+      dns6 = 55,
+      dnsaddr = 56,
+      sctp = 132,
+      udt = 301,
+      utp = 302,
+      unix = 400,
+      ipfs = 421,
+      onion = 444,
+      onion3 = 445,
+      garlic64 = 446,
+      quic = 460,
+      http = 480,
+      https = 443,
+      ws = 477,
+      wss = 478,
+      p2p_websocket_star = 479,
+      p2p_stardust = 277,
+      p2p_webrtc_star = 275,
+      p2p_webrtc_direct = 276,
+      p2p_circuit = 290,
+    };
+
+    const Code code;
+    const ssize_t size;
+    const std::string_view name;
+  };
+
+  /**
+   * Contains a list of protocols and some accessor methods for it
+   */
+  class ProtocolList {
+   public:
+    /**
+     * The total number of known protocols
+     */
+    static const std::size_t kProtocolsNum = 28;
+
+    /**
+     * Returns a protocol with the corresponding name if it exists, or nullptr
+     * otherwise
+     */
+    static constexpr auto get(std::string_view name) -> Protocol const * {
```

Comment:
why return ptr? please try to use raw pointers as less as possible. Maybe returning value or reference is okay here

Reply:
Returning  a value or a reference is barely applicable, as the protocol with the provided name may not exist, std::optional to a reference seems to be ambiguous for such a simple problem(because when I tried to use it this way I came to reference_wrapper), and outcome::result is not really needed, because there may hardly occur any errors, but the absence of a protocol with the given name. I had strong motivation to use a pointer here, but if you still disagree with it, I can change it, of course.

Reply:
because this method is constexpr